configure.ac

```diff
@@ -1033,39 +1033,100 @@ case "$enable_capabilities" in
+
 AC_ARG_ENABLE([delayacct],
               [AS_HELP_STRING([--enable-delayacct],
-                              [enable Linux delay accounting support; requires pkg-config, libnl-3 and libnl-genl-3 @<:@default=check@:>@])],
+                              [enable Linux delay accounting support; requires libnl-3 and libnl-genl-3 @<:@default=check@:>@])],
               [],
               [enable_delayacct=check])
 case "$enable_delayacct" in
-   no)
+   no|yes)
       ;;
    check)
       if test "$my_htop_platform" != linux; then
          enable_delayacct=no
       elif test "$enable_static" = yes; then
          enable_delayacct=no
-      else
-         old_CFLAGS="$CFLAGS"
-         CFLAGS="$CFLAGS -I/usr/include/libnl3"
-         AC_CHECK_HEADERS([netlink/attr.h netlink/handlers.h netlink/msg.h], [enable_delayacct=yes], [enable_delayacct=no])
-         CFLAGS="$old_CFLAGS"
       fi
       ;;
-   yes)
-      old_CFLAGS="$CFLAGS"
-      CFLAGS="$CFLAGS -I/usr/include/libnl3"
-      AC_CHECK_HEADERS([netlink/attr.h netlink/handlers.h netlink/msg.h], [], [AC_MSG_ERROR([can not find required header files netlink/attr.h, netlink/handlers.h, netlink/msg.h])])
-      CFLAGS="$old_CFLAGS"
-      ;;
    *)
       AC_MSG_ERROR([bad value '$enable_delayacct' for --enable-delayacct])
       ;;
 esac
+
+case "$enable_delayacct" in
+   check|yes)
+      m4_ifdef([PKG_PROG_PKG_CONFIG], [
+         PKG_CHECK_MODULES(LIBNL3, libnl-3.0, [], [LIBNL3_CFLAGS="-I/usr/include/libnl3"])
+      ], [
+         LIBNL3_CFLAGS="-I/usr/include/libnl3"
+      ])
+
+      htop_save_CFLAGS=$CFLAGS
+      # New include path searched after what user has specified
+      CFLAGS="$CFLAGS $LIBNL3_CFLAGS"
+      AC_CHECK_HEADERS([netlink/attr.h netlink/handlers.h netlink/msg.h],
+         [],
+         [if test "$enable_delayacct" = yes; then
+            AC_MSG_ERROR([can not find required header files netlink/attr.h, netlink/handlers.h, netlink/msg.h])
+         fi
+         enable_delayacct=no
+         break])
+      CFLAGS=$htop_save_CFLAGS
+
+      if test "$enable_delayacct" != no; then
+         AC_MSG_CHECKING([the search path of libnl-3])
+
+         htop_libnl3_link_succeed=no
+         htop_try_link_libnl3 "$LIBNL3_LIBDIR"
+         if test "$htop_libnl3_link_succeed" = no && test "x$LIBNL3_LIBDIR" = x && test "x$PKG_CONFIG" != x; then
+            LIBNL3_LIBDIR=`$PKG_CONFIG --variable=libdir libnl-3.0 2>/dev/null`
+            if test "x$LIBNL3_LIBDIR" != x; then
+               htop_try_link_libnl3 "$LIBNL3_LIBDIR"
+            fi
+         fi
+
+         if test "x$LIBNL3_LIBDIR" = x; then
+            AC_MSG_RESULT([(default)])
+         else
+            # The path must end with a slash
+            LIBNL3_LIBDIR=`echo "x$LIBNL3_LIBDIR" | sed 's/^x//; s|/*$|/|'`
+            AC_MSG_RESULT([$LIBNL3_LIBDIR])
+         fi
+         AC_DEFINE_UNQUOTED([LIBNL3_LIBDIR], ["$LIBNL3_LIBDIR"], [libnl-3 search path; use the default search paths if empty])
```

Comment:
Is this the right way to define this macro? What about path's like `/usr/lib/malicious.so"); (void)StuffToFixSyntax("`?

Reply:
@BenBE Perhaps forbidding the quote characters (`"`) can help. Technically this is not our bug - It would be the builder that injects the vulnerability and not we that do.

Reply:
It is mostly a matter of proper escaping of variables. It should not fail to build, just because I have my libnl3 in `/usr/local/"special builds"/libs` … ;-)

Reply:
@BenBE That is an undefined behavior AFAICT. At least, it's impossible to specify a path with spaces in an `-L` option to a `LDFLAGS` variable. You _may_ pass a path with spaces in an `-L` option when you run a linker directly, but the makefile format was never designed with that use case. 🙁

Paths with embedded quotation marks would be worse - variable expansions in makefiles would never work with those. The makefile format didn't come with a secure variable expansion by design, unlike variable expansion in shells. Thus if you are working with those paths, you would need the hell a lot of code hacking to make it work (at least you would modify `config.h` manually).